internal/factsengine/gatherers/saphostctrl.go

```diff
@@ -0,0 +1,176 @@
+package gatherers
+
+import (
+	"regexp"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/trento-project/agent/pkg/factsengine/entities"
+	"github.com/trento-project/agent/pkg/utils"
+)
+
+const (
+	SapHostCtrlGathererName               = "saphostctrl"
+	saphostCtrlListInstancesParsingRegexp = `^\s+Inst Info\s*:\s*([^-]+?)\s*-\s*(\d+)\s*-\s*([^,]+?)` +
+		`\s*-\s*(\d+),\s*patch\s*(\d+),\s*changelist\s*(\d+)$`
+	saphostCtrlPingParsingRegexp = `(SUCCESS|FAILED) \( *(\d+) usec\)`
+)
+
+// nolint:gochecknoglobals
+var whitelistedWebmethods = map[string]func(utils.CommandExecutor) (entities.FactValue, *entities.FactGatheringError){
+	"Ping":          handlePing,
+	"ListInstances": handleListInstances,
+}
+
+// nolint:gochecknoglobals
+var (
+	SapHostCtrlCommandError = entities.FactGatheringError{
+		Type:    "saphostctrl-cmd-error",
+		Message: "error executing saphostctrl command",
+	}
+	SapHostCtrlUnsupportedFunction = entities.FactGatheringError{
+		Type:    "saphostctrl-webmethod-error",
+		Message: "requested webmethod not whitelisted",
+	}
+	SapHostCtrlParseError = entities.FactGatheringError{
+		Type:    "saphostctrl-parse-error",
+		Message: "error while parsing saphostctrl output",
+	}
+)
+
+type SapHostCtrlGatherer struct {
+	executor utils.CommandExecutor
+}
+
+func NewDefaultSapHostCtrlGatherer() *SapHostCtrlGatherer {
+	return NewSapHostCtrlGatherer(utils.Executor{})
+}
+
+func NewSapHostCtrlGatherer(executor utils.CommandExecutor) *SapHostCtrlGatherer {
+	return &SapHostCtrlGatherer{
+		executor: executor,
+	}
+}
+
+func (g *SapHostCtrlGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	facts := []entities.Fact{}
+	log.Infof("Starting saphostctrl facts gathering process")
+
+	for _, factReq := range factsRequests {
+		var fact entities.Fact
+
+		if factValue, err := handleWebmethod(g.executor, factReq.Argument); err != nil {
+			fact = entities.NewFactGatheredWithError(factReq, err)
+		} else {
+			fact = entities.NewFactGatheredWithRequest(factReq, factValue)
+		}
+
+		facts = append(facts, fact)
+	}
+
+	log.Infof("Requested %s facts gathered", SapHostCtrlGathererName)
+	return facts, nil
+}
+
+func isWhitelisted(command string) bool {
+	_, ok := whitelistedWebmethods[command]
+	return ok
+}
+
+func handleWebmethod(
+	executor utils.CommandExecutor,
+	incomingCommand string,
+) (entities.FactValue, *entities.FactGatheringError) {
+	if !isWhitelisted(incomingCommand) {
+		gatheringError := SapHostCtrlUnsupportedFunction.Wrap(incomingCommand)
+		log.Error(gatheringError)
+		return nil, gatheringError
+	}
+
+	return whitelistedWebmethods[incomingCommand](executor)
+}
+
+func handlePing(executor utils.CommandExecutor) (entities.FactValue, *entities.FactGatheringError) {
+	saphostctlOutput, commandError := executeSapHostCtrlCommand(executor, "Ping")
+	if commandError != nil {
+		return nil, commandError
+	}
+
+	parsedOutput, parsingError := parsePing(saphostctlOutput)
```

Comment:
All of this is not needed. You should better do `return parsePing(saphostctlOutput)`.
Besides that, I don't personally like to have so many function (handle and parse).
Why don't you put the parsing code here?
It is not used anywhere. I makes reading more difficult as you need to scroll up and down many times to find the code.
I would put everything about the `Ping` here.
The same goes for the other commands